### ISSUE TITLE
Decompress results differ from quicklz

### ISSUE THREAD

**Issue by Thell** [ReSpeak/quicklz#5]
Hello. I've been decompressing about 800 files weekly and was playing with trying to use your rust implementation of quicklz.

~~I altered the `new()` method to use a different key but everything else is the same.~~ [that isn't involved with this, sorry]

81 of these files comes out with differing bytes. All are compressed lvl 3 and the differing bytes aren't scattered about but are in chunks (if that means anything). No errors are being returned and my rust code to call is simply...

```rust
fn decompress(data: &[u8], size: u32) -> Result<Vec<u8>, std::io::Error> {
    let mut r = std::io::Cursor::new(data.as_ref());
    match quicklz::decompress(&mut r, size) {
        Ok(r) => Ok(r),
        Err(e) => {
            eprintln!("{}", e);
            Err(std::io::Error::new(
                std::io::ErrorKind::InvalidData,
                "quicklz error",
            ))
        }
    }
}
```

I did notice that your implementation is not a "full" implementation but rather customized to suit your needs but I was wondering if you might have some input on this.

Attached is a sample of a file that was compressed with some embedded c++ version of quicklz. I can successfully decompress it with c++ code and python (cython) code via the pyquicklz package.

Also attached is the known 'good' decompressed file for comparison.

(renamed from `.bin` to `.txt` to attach...)
[compressed.txt](https://github.com/ReSpeak/quicklz/files/7900851/compressed.txt)
[decompressed.txt](https://github.com/ReSpeak/quicklz/files/7900854/decompressed.txt)

**Comment by Thell** [comment]
This is the smallest of my failing source files and a simple compress/decompress test.

```rust
use std::io::Cursor;
use std::fs::read;
use quicklz::*;

fn main() -> Result<(), std::io::Error> {
    let data = read("data.txt")?;
    let lvl3 = compress(data.as_ref(), CompressionLevel::Lvl3);
    let dec3 = decompress(&mut Cursor::new(&lvl3), data.len() as u32).unwrap();
    assert_eq!(dec3, data);
    Ok(())
}
```

[data.txt](https://github.com/ReSpeak/quicklz/files/7927239/data.txt)

**Comment by Flakebi** [comment]
Hi, thanks for the detailed report and the example files. I found two bugs, one in the decompressor, where it copied a chunk from the wrong offset, and one in the compressor, where it cut off high bits of the offset.
With the fix in bf183250d38d23227925e0c7ebb45dbbb2aafe19, both files are now compressing/decompressing fine. I’ll release an update with the fixes in the next weeks.

**Comment by Thell** [comment]
Tested against my full set of files and all pass.
Thanks for the fix!

**Comment by Flakebi** [comment]
Nice, thanks for testing!

**Comment by Thell** [comment]
Not sure if you want to reopen this or have me make a new issue but I extended my test set and identified another issue.
The following file is being identified as neither lvl 1 or lvl3 in your quicklz but when using c++ quicklz with `#define QLZ_COMPRESSION_LEVEL 3` the file decompresses properly.

`sz_compressed = 352`
`sz_original = 1024`
`compression_level = 3`


[compressed.txt](https://github.com/ReSpeak/quicklz/files/7989692/compressed.txt)
[decompressed.txt](https://github.com/ReSpeak/quicklz/files/7989697/decompressed.txt)

**Comment by Thell** [comment]
Hello again, I was just wondering about the disposition of this issue.

**Comment by Flakebi** [comment]
Hi, sorry for the late response, I didn’t have much time to look at this.
It looks like compressing `decompressed.txt` with the rust quicklz returns a different result than your c++ `compressed.txt`, but I have no problems decompressing both of them with the rust implementation.
The following runs through fine for me:
```rust
        let dec_should = include_bytes!("../decompressed.txt");
        let comp_should = include_bytes!("../compressed.txt");
        let dec_is =
            decompress(&mut Cursor::new(comp_should.as_slice()), 10000)
                .unwrap();
        assert_eq!(&dec_should[..], &dec_is);

        let comp = compress(dec_should, CompressionLevel::Lvl3);
        let mut r = Cursor::new(comp.as_slice());
        let dec = decompress(&mut r, 1000000).unwrap();
        assert_eq!(&dec_should[..], &dec);

        // This fails, rust quicklz returns a different compressed result
        // assert_eq!(comp.as_slice(), &comp_should[..]);
```

**Comment by Thell** [comment]
Thanks for looking. It sure should work if you compress and then recompress... if it didn't that'd be a big issue, lol. That's true here as well.

It seems the original bytes I receive don't have a header **but** _that does not hinder the official quicklz implementation_. Since the official code is what I am using (c code, compiled with c++) and all compressed files process and output decompressed contents as expected using that code I'm guessing that is what we'd want a rust implementation to do as well.

Notice in the output below the recompressed bytes look to include a header; all other bytes are the same but for some reason the rust decompress is not identifying the compression properly and is throwing the compression level panic.

```
>>> origbytes
b'o_\x01\x00\x00\x00\x04\x00\x00\x00\x00\x03\x80o\xb7n\xcb\xc9\xe1\x9c\x8eA\xefXj\xf7\xcaO\x0e\x03~\x04\x00\x037\x04\x00\x14b\x01\x82\x07(\xeedeh\x9aOW\x080\x00\x80\x05\xbe\r\x83\x1e\x04\x00\xf1!.$c\xff\xe7\x8a\x83*0\x00\x832\x04\x00\x1f\x9c(\x01\xc4\x08\x1b\xca\xe8\\.\xb0\xde\x82\xf9\x02\xe3\x00\x00\x80\x80\xfa\xa8\x95\xbe\\`\xdf.\x85\x07\xe2\xdd27\x8df\xe3]f\x7f3\xab\x88\x16\x08\x86\x1d\xd2U\x9ed\xf7\x02\x04\x08\x904\x16\x04\xa9\x91C\xcc\x07\x0bW\xdf\x16\x04\x7f\xac\x9f#\x19\xa2a\xda\x16\x04H\xb9{\xf9\xeat\xd2\xc4\x16\x04\x1d\xd5@\x80\x00\x81\x0e\xb9\x94X\xfe\xb7\x16\x04z\xbcH\x1a\xc6\xb9\xe0z\x16\x04\xff\xd9\xdb\xde\x17\xeb\xa4\x9f\x16\x04\x14\xeeS\xd4\xc9|\x04\x08\x10\xa0\x03{\x16\x04M\xbdx2\xbc\xfb\xd9\xa1\x16\x04 \xd8\x1a\xb7a\x04\xbab\x16\x043\xf4}\xca\x02\xafu\xee\x16\x04\x88\x80\x00\x01\x82\tc\xe0\r\xc92\x8c\x16\x04\xe2\xe1\xd9\xc0\x05E\x97\xd2\x16\x04h\x14H\x0e\xe6\x88\xcf\xc7\x16\x04\n\x10\x1f?\xcc\x08\x10\x00\x80\\\x18\x83\x16\x04F\xcbz\xc6\x1f\x9c\x1eq\x16\x04X\xd9\xf2\xcfD+\xd7\xaci\x1b\xf1i\xa3\xbb\xff]\x8aU@\x80\x00\x80\xd5\x0e\t\xc3\x85(\x16\x08\x90E!\\\xba\xefz\x19\x16\x04A\xefXj\xf7\xcaO\x0e\x00'
>>> recompressedbytes
b'Ni\x01\x00\x00`\x01\x00\x00o_\x01\x00\x00\x00\x04\x00\x00\x00\x00\x03\x80o\xb7n\xcb\xc9\xe1\x9c\x8eA\xefXj\xf7\xcaO\x0e\x03~\x04\x00\x037\x04\x00\x14b\x01\x82\x07(\xeedeh\x9aOW\x080\x00\x80\x05\xbe\r\x83\x1e\x04\x00\xf1!.$c\xff\xe7\x8a\x83*0\x00\x832\x04\x00\x1f\x9c(\x01\xc4\x08\x1b\xca\xe8\\.\xb0\xde\x82\xf9\x02\xe3\x00\x00\x80\x80\xfa\xa8\x95\xbe\\`\xdf.\x85\x07\xe2\xdd27\x8df\xe3]f\x7f3\xab\x88\x16\x08\x86\x1d\xd2U\x9ed\xf7\x02\x04\x08\x904\x16\x04\xa9\x91C\xcc\x07\x0bW\xdf\x16\x04\x7f\xac\x9f#\x19\xa2a\xda\x16\x04H\xb9{\xf9\xeat\xd2\xc4\x16\x04\x1d\xd5@\x80\x00\x81\x0e\xb9\x94X\xfe\xb7\x16\x04z\xbcH\x1a\xc6\xb9\xe0z\x16\x04\xff\xd9\xdb\xde\x17\xeb\xa4\x9f\x16\x04\x14\xeeS\xd4\xc9|\x04\x08\x10\xa0\x03{\x16\x04M\xbdx2\xbc\xfb\xd9\xa1\x16\x04 \xd8\x1a\xb7a\x04\xbab\x16\x043\xf4}\xca\x02\xafu\xee\x16\x04\x88\x80\x00\x01\x82\tc\xe0\r\xc92\x8c\x16\x04\xe2\xe1\xd9\xc0\x05E\x97\xd2\x16\x04h\x14H\x0e\xe6\x88\xcf\xc7\x16\x04\n\x10\x1f?\xcc\x08\x10\x00\x80\\\x18\x83\x16\x04F\xcbz\xc6\x1f\x9c\x1eq\x16\x04X\xd9\xf2\xcfD+\xd7\xaci\x1b\xf1i\xa3\xbb\xff]\x8aU@\x80\x00\x80\xd5\x0e\t\xc3\x85(\x16\x08\x90E!\\\xba\xefz\x19\x16\x04A\xefXj\xf7\xcaO\x0e\x00'  
>>> origbytes == recompressedbytes[9:]
True
```

**Comment by Flakebi** [comment]
I still have problems reproducing your issue.

I compressed the `decompressed.txt` that you attached with the rust quicklz and with the C quicklz. Both result in the exact same compressed data.
The result I got is slightly different from the `compressed.txt` you attached, it starts with 0x4f (`O_\x01`) instead of 0x6f, so your C quicklz library has streaming mode enabled? IIRC this rust implementation does not support streaming mode and will simply ignore these bits in the header (there also is an extra `0` byte at the end of your `compressed.txt`, I assume because of a C string, both implementations should stop reading the file before this byte, so it doesn’t matter).

In this case, the streaming setting doesn’t seem to make a difference and decompressing with the rust and C quicklz gives the same decompressed result, which is the same as your `decompressed.txt` :)

**Comment by Thell** [comment]
Thank you, again, for taking another look at this. I _must_ be messing something up on my interface then.
I will check it again and update/close this, hopefully, tomorrow.

**Comment by Thell** [comment]
Man oh man... it took a long time to find those erroring files again but I did find them.

I ran the extractions with the c implementation and all extracted fine (59000+ files) and then ran with the rust implementation and got 125 failures. While putting together the files to post a reply explaining what was being observed I re-ran all of the intermediate (decrypted) files from the c solution through the rust solution and they worked! This obviously pointed to a problem on my end and now, with a place to look, I found the logic error (an `.unwrap_or(false)` where an `.unwrap_or(true)` should have been. 🤦 Thank you so much for being patient and looking at those files. I'm pretty excited to continue working on this now. 😄
